pennylane/devices/qubit/preprocess.py

```diff
@@ -77,6 +78,76 @@ def _operator_decomposition_gen(
 #######################
 
 
+def validate_and_expand_adjoint(
+    circuit: qml.tape.QuantumScript,
+) -> qml.tape.QuantumScript:  # pylint: disable=protected-access
+    """Function for validating that the operations and observables present in the input circuit
+    are valid for adjoint differentiation.
+
+    Args:
+        circuit(.QuantumTape): the tape to validate
+
+    Returns:
+        .QuantumTape: The expanded tape, such that it is supported by adjoint differentiation
+
+    Raises:
+        QuantumFunctionError: if the circuit is invalid for adjoint differentiation
+    """
+    # Check validity of measurements
+    measurements = []
+    for m in circuit.measurements:
+        if not isinstance(m, ExpectationMP):
+            raise qml.QuantumFunctionError(
+                "Adjoint differentiation method does not support"
+                f" measurement {m.__class__.__name__}"
+            )
+
+        if m.obs.name in {"Hamiltonian"}:
+            raise qml.QuantumFunctionError(
+                f"Adjoint differentiation method does not support observable {m.obs.name}."
+            )
+
+        if not hasattr(m.obs, "base_name"):
+            m.obs.base_name = None  # This is needed for when the observable is a tensor product
+
+        measurements.append(m)
+
+    expanded_ops = []
+    for op in circuit._ops:
+        if op.num_params > 1:
+            if not isinstance(op, qml.Rot):
+                raise qml.QuantumFunctionError(
+                    f"The {op.name} operation is not supported using "
+                    'the "adjoint" differentiation method'
+                )
+            ops = op.decomposition()
+            expanded_ops.extend(ops)
+        elif not isinstance(op, qml.operation.StatePrep):
```

Comment:
do we expect StatePrep ops to be in the `_ops` category?

Reply:
Yep. The nice thing about this validation is that even is a user puts it later in the list of operations, it won't be appended to the expanded tape so adjoint diff will work fine.

Reply:
Not really. And if they are, they probably won't have a matrix, which means the standard preprocessing will expand them away.